src/cocaine-app/sched/__init__.py

```diff
@@ -0,0 +1,417 @@
+                logger.info('Cannot stop job {} type {} since it has >= priority and no force'.format(job.id, job.type))
+                continue
+
+            if job.status in (jobs.Job.STATUS_NOT_APPROVED, jobs.Job.STATUS_NEW):
+                cancellable_jobs.append(job.id)
+                continue
+
+            if job.type not in STOP_ALLOWED_TYPES:
+                logger.info('Cannot stop job {} of type is {}'.format(job.id, job.type))
+                continue
+
+            cancellable_jobs.append(job.id)
+
+        if len(cancellable_jobs) == 0:
+            logger.info("No jobs to cancel while {} are blocking".format(len(job_ids)))
+            return False
+
+        logger.info("Analyzing demand {} while cancellable_jobs are {}".format(demand, cancellable_jobs))
+
+        # Analyze whether stopping of cancelable_jobs would be enough
+        for res in demand:
+            res_demand = demand[res]  # how much of resource of this type and id, the job wants to consume
+            usage_if_cancel = 0  # how much of resource would be still consumed when we'd cancel all cancellable jobs
+
+            for j in self.res[res]:
+                if j[1] not in cancellable_jobs:
+                    usage_if_cancel += j[0]
+
+            if res_demand > (100 - usage_if_cancel):
+                logger.info("No sense to cancel cause of res={} (demand={} < usage_if_cancel={}, jobs are {})".format(
+                    res, res_demand, usage_if_cancel, self.res[res]))
+                return False
+
+        # XXX: Now we cancel all jobs while we may be satisfied with less amount. Fix that
+
+        try:
+            self.job_processor.stop_jobs_list(filter(lambda x: x.id in cancellable_jobs, existing_jobs))
+        except:
+            logger.exception("Failed to cancel jobs {}".format(cancellable_jobs))
+            return False
+
+        logger.info("Successfully cancelled {}".format(cancellable_jobs))
+        return True
+
+    def process_jobs(self, job_type, jobs_param_list, sched_params, filter_func=None):
+        """
+        Scheduler creates jobs with specified params of specified type. Sched_params and filter_func
+        :param job_type: job type
+        :param jobs_param_list: a list of dictionaries with params for jobs created.
+                                A number of job created <= len(jobs_param_list)
+        :param sched_params: in a common case sched params are stored in config in "scheduler/{starter}" section.
+                            But in certain cases the starter may want to redefine some parameters.
+                            For example, one may want to increase priority or decrease max_deadline_time
+        :param filter_func: Filter function are called before job creation to extra-check whether those jobs are needed
+                            See more details on wiki
+                            (https://wiki.yandex-team.ru/users/aniam/mm-scheduler/#generacijadzhobovstartdzhobov)
```

Comment:
Done in https://github.com/yandex/mastermind/pull/417/commits/e7a0f53468278b6dc781abea9c82cb3cb7350f4e